README.md
# Large file (vsc-lfs) README

[![Version](https://vsmarketplacebadge.apphb.com/version/mbehr1.vsc-lfs.svg)](https://marketplace.visualstudio.com/items?itemName=mbehr1.vsc-lfs)

This extensions allows you to (in a sane way) work around the restriction described as vscode issue [#27100](https://github.com/microsoft/vscode/issues/27100) and feature request [#31078](https://github.com/microsoft/vscode/issues/31078).
So you can open large files (e.g. a few hundred MBs) and they get passed to the extension host / to your extensions.

**Note:** Use with care! The vs-code authors decided to impose the size restriction for a good reason. Use it only when you do really need it and not for every file. 

## Features

- open large (>50MB) files and still allow your other extensions to process it.
- replacement features allows to search and replace file content at opening. See below for setting `vsc-lfs.replacements`.

## Usage

Use the command "open large file..." and select the file you want in the following dialog. The dialog contains a file filter that can be customized (see settings).

## Extension Settings

This extension contributes the following settings:

* `vsc-lfs.fileFilters`: Array of strings that contain the file filters to apply for the open large file dialog. Default to ["txt", "TXT", "log", LOG"].
* `vsc-lfs.reReadTimeout`: Time after which a re-read takes place. If your file doesn't open try to increase this value. Defaults to 5s.
* `vsc-lfs.replacements`: Optional array of objects with name and filters property. Filters is a array of objects with search and replace properties. If replacements are defined at *open large file...* command you can select any of the replacements. The selected ones will be applied as search and replace line-by-line. Search and replace must not increase the file size but the intention is to shrink the file. search is a javascript regular expression and replace can contain special replacement patterns like *$&* or *$n*. For details see [Specifying a string as a parameter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace#Specifying_a_string_as_a_parameter) E.g.
```
[{
  "name": "shorten line and date",
  "filters": [
    {
      "search": "line: (\\d*)$", // regex to search lines ending with 'line: <number>'. Capturing the number.
      "replace": "l:$1" // replace with l:<number>
    },
    { // replace all ' date:' with ' d:'
      "search": " date:",
      "replace": " d:"
    }
  ]
},
{
  "name": "delete all debug lines",
  "filters": [
    "search": "^.*\\[DEBUG\\].*$", // full line with [DEBUG]. [ ] need to be escaped in reg exp.
    "replace": "" // replace with empty line
  ]
}]
```
**Note:**  take care to escape \ properly in json as \\\\.

**Note:**  currently the search/replace is applied line by line.

## Known Issues

* Currently the files are opened read-only even if they are not from the original filesystem.

## Contributions

Any and all test, code or feedback contributions are welcome.
Open an [issue](https://github.com/mbehr1/vsc-lfs/issues) or create a pull request to make this extension work better for all.

[![Donations](https://www.paypalobjects.com/en_US/DK/i/btn/btn_donateCC_LG.gif)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=2ZNMJP5P43QQN&source=url) Donations are welcome! (Contact me for commercial use or different [license](https://creativecommons.org/licenses/by-nc-sa/4.0/legalcode)).

[GitHub ♥︎ Sponsors are welcome!](https://github.com/sponsors/mbehr1)

## Planned features

* Add *autoApplyIf* option to replacements for e.g. above a certain file size or file name patterns.
* Remove readonly restriction.
* Investigate: seems like file changes are already supported. VS Code seems to understand our uri format and triggers a readFile on file change.
  Support watch for removing/deleted files.
* Add debouncing / delayed updates features to limit the amount of file changes reflected to extensions.

## Release Notes

See [Changelog](./CHANGELOG.md)


CHANGELOG.md
# Change Log for 'vsc-lfs':

### [1.4.1](https://github.com/mbehr1/vsc-lfs/compare/v1.4.0...v1.4.1) (2021-05-16)


### Bug Fixes

* fileFilters not declared as array of strings ([e833205](https://github.com/mbehr1/vsc-lfs/commit/e833205238598f557cc6eba0e069bcf3fff14078))

## [1.4.0](https://github.com/mbehr1/vsc-lfs/compare/v1.3.6...v1.4.0) (2021-05-13)


### Features

* limit support of untrusted workspaces ([14f71c8](https://github.com/mbehr1/vsc-lfs/commit/14f71c89cb29ea314cb9ee32af7cb371737ac96d))

### [1.3.6](https://github.com/mbehr1/vsc-lfs/compare/v1.3.5...v1.3.6) (2021-02-06)


### Bug Fixes

* issue links in readme ([04467a2](https://github.com/mbehr1/vsc-lfs/commit/04467a2aa1eab7ef07f7e2ab8f8970209cf2eabd))

### [1.3.5](https://github.com/mbehr1/vsc-lfs/compare/v1.3.4...v1.3.5) (2020-12-27)

### [1.3.4](https://github.com/mbehr1/vsc-lfs/compare/v1.3.3...v1.3.4) (2020-12-27)


### Bug Fixes

* **changelog:** changelogs not updated ([a010dcb](https://github.com/mbehr1/vsc-lfs/commit/a010dcb5fd73cd6ef9ef025f689e65b7b24371e8))

### [1.3.3]
* **CI/build:** add automatic generation of vsix package.

### [1.3.2](https://github.com/mbehr1/vsc-lfs/compare/v1.3.1...v1.3.2) (2020-12-27)


### Bug Fixes

* **copyright:** update copyright for 2021 ([7352d79](https://github.com/mbehr1/vsc-lfs/commit/7352d7908cee19e66afee78f1190bf3d7a7d57d5))

### [1.3.1]
- updated README example

### [1.3.0]
- added *replacements* configuration option for search/replace feature to shrink the file content on loading. This can be used to e.g. replace unneeded content in large files before loading.

### [1.2.1]
- update dependencies after github security advisory
- fix extensionVersion

### [1.2.0]
- changed watch dispose behaviour to reduce number of re-loads on changing active document.
- implemented readDirectory so that breadcrumb navigation is working.

### [1.1.5]
- fixed a bug with wrongly detected directories as files.

### [1.1.4]
- fixed a bug with small files returning wrong size on stat.

### [1.1.3]
- add support for automatic reopen at startup

### [1.1.2]
- Set default re-read timeout to 5s.

### [1.1.1]
- Added config for re-read timeout: `vsc-lfs.reReadTimeout`.

### [1.1.0]

- Added telemetry using vscode-extension-telemetry with events: 'activate' and 'open large file' (measurement: fileSize). The telemetry is following the user setting: telemetry.enableTelemetry.
### [1.0.2]
- Initial release


src/test/suite/lfsProvider.test.ts
import * as assert from 'assert';

import * as vscode from 'vscode';
import * as myExtension from '../../extension';

function sleep(ms: number): Promise<void> {
    return new Promise(resolve => {
        setTimeout(resolve, ms);
    });
}

suite('LFSProvider class Test Suite', () => {
    vscode.window.showInformationMessage('Start all LFSProvider class tests');
    let lfsP: myExtension.LFSProvider = new myExtension.LFSProvider(undefined);
    lfsP.limitedSize = 10;
    lfsP.reReadTimeout = 50; // 0.05s for test

    test('first stat limited in size', async () => {
        console.log(`cwd=${__dirname}`);
        const smallFileUri = vscode.Uri.parse(`lfs:${__dirname.replace("out/", "src/")}/../smallFile.txt`);
        // first one should be limited in size
        assert.equal(10, lfsP.stat(smallFileUri).size);
        // next ones as well until readFile happens
        assert.equal(10, lfsP.stat(smallFileUri).size);
        assert.equal(10, (await lfsP.readFile(smallFileUri)).length);
        // wait for reread...
        await sleep(60);
        // now full size:
        assert.equal(60, lfsP.stat(smallFileUri).size);
        assert.equal(60, (await lfsP.readFile(smallFileUri)).length);
    });

    test('access not existing file', () => {
        const uri = vscode.Uri.parse("lfs:/nonExistitingFile.txt");
        try {
            lfsP.stat(uri);
            assert.fail("Expected an error");
        } catch (err) {
        };

        assert.throws(() => {
            lfsP.stat(uri);
        }, Error, "no error thrown");
    });
});


src/extension.ts
/* --------------------
 * Copyright (C) Matthias Behr, 2020 - 2021
 * 
 * This extension provides a temporary solution until vscode issue #27100, feature request #31078
 * is changed/implemented.
 * 
 * Missing features:
 * [ ] support watch for changing files
 * [ ] remove readonly restriction
 */

import * as vscode from 'vscode';
import * as fs from 'fs';
import TelemetryReporter from 'vscode-extension-telemetry';
import { createInterface } from 'readline';
const { once } = require('events');

let reporter: TelemetryReporter;

export function activate(context: vscode.ExtensionContext) {
	console.log('vsc-lfs.activated. Use "open large file" command to open a large file :-)');

	const extensionId = 'mbehr1.vsc-lfs';
	const extension = vscode.extensions.getExtension(extensionId);

	if (extension) {
		const extensionVersion = extension.packageJSON.version;

		// the aik is not really sec_ret. but lets avoid bo_ts finding it too easy:
		const strKE = 'ZjJlMDA4NTQtNmU5NC00ZDVlLTkxNDAtOGFiNmIzNTllODBi';
		const strK = Buffer.from(strKE, "base64").toString();
		reporter = new TelemetryReporter(extensionId, extensionVersion, strK);
		context.subscriptions.push(reporter);
		reporter?.sendTelemetryEvent('activate');
	} else {
		console.log("vsc-lfs: not found as extension!");
	}

	const lfsP = new LFSProvider(context.globalState);
	context.subscriptions.push(vscode.workspace.registerFileSystemProvider('lfs', lfsP, { isReadonly: true, isCaseSensitive: true }));

	context.subscriptions.push(vscode.commands.registerCommand('extension.lfsOpenFile', async () => {
		// eslint-disable-next-line @typescript-eslint/naming-convention
		return vscode.window.showOpenDialog({ canSelectFiles: true, canSelectFolders: false, canSelectMany: false, filters: { 'Large files': <Array<string>>(vscode.workspace.getConfiguration().get("vsc-lfs.fileFilters")) }, openLabel: 'Select large file to open...' }).then(
			async (uris: vscode.Uri[] | undefined) => {
				if (uris) {
					uris.forEach(async (uri) => {
						const fileStat = fs.statSync(uri.fsPath);
						console.log(`open large file with size=${fileStat.size} from URI=${uri.toString()}`);
						let lfsUri = uri.with({ scheme: 'lfs' });

						// any replacements defined?
						// if so ask user to select. todo add autoApplyIf option
						const regexs: { search: string, replace: string }[] = [];
						const replacements = <Array<{ name: string, filters: any[] }> | undefined>(vscode.workspace.getConfiguration().get("vsc-lfs.replacements"));
						if (replacements?.length) {
							interface ReplacementQuickPick {
								label: string,
								picked?: boolean,
								data: { name: string, filters: any[] }
							}
							const selected = await vscode.window.showQuickPick<ReplacementQuickPick>(replacements.map(r => { return { label: r.name, data: r }; }), { canPickMany: true });
							if (selected && selected.length) {
								//console.log(` selected='${JSON.stringify(selected)}'`);
								for (const sel of selected) {
									sel.data.filters.forEach((f) => {
										regexs.push(f);
									});
								}
							}
						}
						lfsP.markLimitSize(lfsUri, true, regexs);
						vscode.workspace.openTextDocument(lfsUri).then((value) => { vscode.window.showTextDocument(value, { preview: false }); });
					});
				}
			}
		);
	}));

	context.subscriptions.push(vscode.workspace.onDidCloseTextDocument((doc) => {
		lfsP.onDidCloseTextDocument(doc.uri);
	}));

}

export class LFSProvider implements vscode.FileSystemProvider { // export only for test access?

	limitedSize: number = 1024 * 1024; // limit to 1MB on first read.
	reReadTimeout: number = vscode.workspace.getConfiguration().get<number>('vsc-lfs.reReadTimeout') ?
		<number>(vscode.workspace.getConfiguration().get<number>('vsc-lfs.reReadTimeout')) : 5000; // 5s default
	private _fileFilters: string[] = <Array<string>>(vscode.workspace.getConfiguration().get("vsc-lfs.fileFilters"));

	private _uriMap: Map<string, { limitSize: boolean, fileBuffer?: Buffer, filters?: { search: string, replace: string }[] }> = new Map<string, { limitSize: boolean, fileBuffer?: Buffer, filters?: { search: string, replace: string }[] }>();

	constructor(private storage?: vscode.Memento) {

	}

	markLimitSize(uri: vscode.Uri, limitSize = true, filters?: { search: string, replace: string }[]) {
		this.storage?.update(`filters_${uri.toString()}`, filters);
		if (!this._uriMap.has(uri.toString())) {
			console.log(`vsc-lfs.markLimitSize(uri=${uri.toString()})... limitSize=${limitSize}`);
			this._uriMap.set(uri.toString(), { limitSize: limitSize, filters: filters });
		} else {
			let curSet = this._uriMap.get(uri.toString());
			if (curSet) {
				curSet.limitSize = limitSize;
				curSet.filters = filters;
				this._uriMap.set(uri.toString(), curSet);
			}
		}
	}

	/**
	 * read a file line by line and apply replaces per line
	 * filters are only allowed to shrink the file!
	 * @param fsPath filepath of the file to read
	 * @param filters array of search/replace to be applied per line
	 * @returns Buffer with the full replaced file content 
	 */
	async readWithFilters(fsPath: string, filtersJson: { search: string, replace: string }[]): Promise<Buffer | undefined> {
		try {
			const fileStat = fs.statSync(fsPath);
			const buf = Buffer.allocUnsafe(fileStat.size);
			let bufUsed = 0;
			const filters = filtersJson.map(f => { return { search: new RegExp(f.search, 'g'), replace: f.replace }; });
			const filtersLen = filters.length;
			try {
				const rl = createInterface({
					input: fs.createReadStream(fsPath),
					crlfDelay: Infinity
				});

				rl.on('line', (line) => {
					// we keep empty lines as empty:
					if (!line.length) {
						bufUsed += buf.write('\n', bufUsed); // todo detect \r\n? 
						return;
					}
					// check each search regExp:
					for (let i = 0; i < filtersLen; ++i) {
						line = line.replace(filters[i].search, filters[i].replace);
						// todo check whether it's faster to always replace line or 
						// e.g. match first and then replace.
					}
					if (line.length) {
						if (bufUsed) {
							bufUsed += buf.write('\n', bufUsed); // todo detect \r\n? 
						}
						bufUsed += buf.write(line, bufUsed);
					}
				});
				await once(rl, 'close');
			}
			catch (err) {
				console.error(`readWithFilters got inner err='${err}'`);
			}
			return buf.slice(0, bufUsed);
		}
		catch (err) {
			console.error(`readWithFilters got outer err='${err}'`);
		}
		return undefined;
	}

	stat(uri: vscode.Uri): vscode.FileStat {

		if (!this._uriMap.has(uri.toString())) {
			this._uriMap.set(uri.toString(), { limitSize: true, filters: this.storage?.get(`filters_${uri.toString()}`) });
		}
		// todo do we have to return here the size of the filtered content?

		const limitSize = this._uriMap.get(uri.toString())?.limitSize;

		// console.log(`vsc-lfs.stat(uri=${uri.toString()})... _limitSize=${limitSize}`);
		const fileUri = uri.with({ scheme: 'file' });
		const realStat = fs.statSync(fileUri.fsPath);
		let fileStat: vscode.FileStat = { ctime: realStat.ctime.valueOf(), mtime: realStat.mtime.valueOf(), size: (limitSize && realStat.size > this.limitedSize) ? this.limitedSize : realStat.size, type: realStat.isFile() ? (vscode.FileType.File) : (realStat.isDirectory() ? vscode.FileType.Directory : vscode.FileType.Unknown) };
		// console.log(` stat returning size=${fileStat.size}/${realStat.size}`);
		return fileStat;
	}

	async readFile(uri: vscode.Uri): Promise<Uint8Array> {
		if (!this._uriMap.has(uri.toString())) {
			this._uriMap.set(uri.toString(), { limitSize: true, filters: this.storage?.get(`filters_${uri.toString()}`) });
		}
		let curSet = this._uriMap.get(uri.toString());
		if (!curSet) {
			throw vscode.FileSystemError.Unavailable();
		}

		console.log(`vsc-lfs.readFile(uri=${uri.toString()})... _limitSize=${curSet.limitSize} filters=${JSON.stringify(curSet.filters)}`);

		if (!curSet.fileBuffer) {
			const fileUri = uri.with({ scheme: 'file' });
			if (curSet.filters === undefined || curSet.filters.length === 0) {
				console.log(` largeFile reading ${fileUri.fsPath} without filters`);
				curSet.fileBuffer = fs.readFileSync(fileUri.fsPath);
			} else {
				// reading with filters
				curSet.fileBuffer = await this.readWithFilters(fileUri.fsPath, curSet.filters);
			}
			if (!curSet.fileBuffer) {
				throw vscode.FileSystemError.FileNotFound();
			}
			reporter?.sendTelemetryEvent('open large file', undefined, { 'fileSize': curSet.fileBuffer.length, 'filters': curSet.filters?.length || 0 });
		}

		if (curSet.limitSize && curSet.fileBuffer && (curSet.fileBuffer.length <= this.limitedSize)) {
			// the file is already smaller than our limit
			curSet.limitSize = false;
		}

		if (curSet.limitSize) {
			setTimeout(() => {
				//console.log(` readFile triggering re-read...`);
				let curSet = this._uriMap.get(uri.toString());
				if (curSet) {
					curSet.limitSize = false;
					this._uriMap.set(uri.toString(), curSet);
				}
				this._emitter.fire([{ type: vscode.FileChangeType.Changed, uri: uri }]);
			}, this.reReadTimeout);

			return curSet.fileBuffer.slice(0, 0 + this.limitedSize);
		} else {
			let toRet = curSet.fileBuffer;
			curSet.fileBuffer = undefined;
			return toRet;
		}
	}

	private _emitter = new vscode.EventEmitter<vscode.FileChangeEvent[]>();
	readonly onDidChangeFile: vscode.Event<vscode.FileChangeEvent[]> = this._emitter.event;

	watch(uri: vscode.Uri): vscode.Disposable {
		console.log(`vsc-lfs.watch(uri=${uri.toString()}...`);
		return new vscode.Disposable(() => {
			console.log(`vsc-lfs.watch.Dispose ${uri}`);
			let curSet = this._uriMap.get(uri.toString());
			if (curSet) {
				// on changing to a different editor the watch dispose is called already.
				// let's try to optimize that.
				// we don't delete the key here as we don't know whether the file is still open.
				//curSet.limitSize = true;
				curSet.fileBuffer = undefined;
				this._uriMap.set(uri.toString(), curSet);
			}
		});
	}

	onDidCloseTextDocument(uri: vscode.Uri) {
		// console.log(`vsc-lfs onDidCloseTextDocument(${uri.toString()})...`);
		let curSet = this._uriMap.get(uri.toString());
		if (curSet) {
			console.log(` onDidCloseTextDocument(${uri.toString()}) closing internally.`);
			curSet.limitSize = true;
			curSet.fileBuffer = undefined;
			this._uriMap.set(uri.toString(), curSet);
		}
	}

	readDirectory(uri: vscode.Uri): [string, vscode.FileType][] {
		console.log(`vsc-lfs.readDirectory(uri=${uri.toString()}...`);
		let entries: [string, vscode.FileType][] = [];
		// list all dirs and files matching our filter:
		const dirEnts = fs.readdirSync(uri.fsPath, { withFileTypes: true });
		for (var i = 0; i < dirEnts.length; ++i) {
			// console.log(` vsc-lfs.readDirectory found ${dirEnts[i].name}`);
			if (dirEnts[i].isDirectory()) {
				entries.push([dirEnts[i].name, vscode.FileType.Directory]);
			} else {
				if (dirEnts[i].isFile()) {
					let matchesFilter = this._fileFilters.length > 0 ? false : true;
					for (let j = 0; j < this._fileFilters.length; ++j) {
						if (dirEnts[i].name.endsWith(this._fileFilters[j])) {
							matchesFilter = true;
							break;
						}
					}
					if (matchesFilter) {
						entries.push([dirEnts[i].name, vscode.FileType.File]);
					}
				}
			}
		}
		return entries;
	}

	writeFile(uri: vscode.Uri, content: Uint8Array, options: { create: boolean, overwrite: boolean }): void {
		console.log(`vsc-lfs.writeFile(uri=${uri.toString()}...`);
		throw vscode.FileSystemError.NoPermissions();
	}

	rename(oldUri: vscode.Uri, newUri: vscode.Uri, options: { overwrite: boolean }): void {
		console.log(`vsc-lfs.rename(oldUri=${oldUri.toString()}...`);
		throw vscode.FileSystemError.NoPermissions();
	}

	delete(uri: vscode.Uri): void {
		console.log(`vsc-lfs.delete(uri=${uri.toString()}...`);
		throw vscode.FileSystemError.NoPermissions();
	}

	createDirectory(uri: vscode.Uri): void {
		console.log(`vsc-lfs.createDirectory(uri=${uri.toString()}...`);
		throw vscode.FileSystemError.NoPermissions();
	}

}
